ARCs/arc-0015.md

```diff
@@ -0,0 +1,254 @@
+---
+arc: 15
+title: Algorand Standard for Encrypted Short Messages 
+description: A standard scheme for encryption/decryption that allows for private messages.
+author: Stéphane Barroso (@sudoweezy), (@urtho)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/*
+status: Draft
+type: Standards Track
+category: ARC
+created: 2021-08-07
+---
+
+# Algorand Standard for Encrypted Short Messages 
+
+## Abstract
+
+Algorand public account addresses (Ed25519) can be converted to DH keys (Curve25519) suitable for encryption.  
+A sender can use a converted public key to encrypt message contents. Transaction signature verification assures the message's authenticity. 
+The recipient can decrypt the message successfully with a derivative of the Algorand private key associated with the public key used for encryption. 
+
+## Specification
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+For the rest of the document, we will use the term *Receiver* to represent the set of accounts (public account ed25519, multisig account, contract account, application account)
+
+> Comments like this are non-normative.
+
+
+### Account's message Application
+
+To receive a message, the *Receiver* **MUST** create an application that follows the following convention:
+
+- A Local State named `public_key` **MUST** containing a public account `Ed25519-Address`
+  - If the *Receiver* is a public account, `public_key` **MAY** be equal to the account's Address as `public_key`, however for security concern it is **NOT RECOMMENDED**.
+  > If `public_key` = *Receiver* `Ed25519-Address`, the decryption will be done by using the private key of the *Receiver* directly. 
+- A Box `inbox` with:
+  - Keys equal to the concatenation of the sender's address, and the round when the message was sent.
+  > example: concat("UAME...RSRI", "_", "24518905")
+  - Values are encoded messages
+ 
+### Extensible message format
+
+The payload of the messages is defined as UINT8Array MessagePack encoding of an **ALGOSMSV0_PAYLOAD** object. 
+
+The **ALGOSMSV0_PAYLOAD** object contains, among other fields, an encrypted UINT8Array MessagePack encoding of **ALGOSMSV0_MESSAGE**.
+
+#### Definitions
+
+**ALGOSMSV0_TYPE_MESSAGE**: `algosms/v0:message`
+
+**ALGOSMSV0_TYPE_MSG**: `algosms/v0:msg`
+
+```ts
+// UINT8Array ecrypted contents as msgpack encoded object
+interface ALGOSMSV0_PAYLOAD {
+  /** Object type eq to ALGOSMSV0_TYPE_MESSAGE * */
+  t: string;
+
+  /** one-time public key of the NACL's sealed box * */
+  otPK: Uint8Array;
+
+  /** cipherText of the ALGOSMSV0_MESSAGE structure * */
+  cT: Uint8Array;
+}
+```
+
+```ts
+// UINT8Array ALGOSMSV0_MESSAGE contents as msgpack encoded object
+interface ALGOSMSV0_MESSAGE {
```

Comment:
Do we want to enforce the message object format?
I think it may be hard to know in advance what will be used.
msgpack may be hard to decode in other chains or in ZK proofs for example.
So why not allowing any arbitrary binary message?

Also, if "from" is added, there is a need for the wallet to confirm that it is not impersonated, which is not straightforward as there are two .algo domain name providers for example.